File: quick_match_chapters.py
# This script automatically matches chapters for audiobooks using AudiobookShelf API.
# Handle with care! If chapters are manually added to a book, they might get overwritten.
# Chapters are considered "missing" based on a threshold setting.
# For example, if there are 123 chapters locally and 127 are found, chapters will be replaced.
# However, if 120-126 chapters are found, they won’t be replaced.

import time
import requests

# Configuration constants
CHAPTER_THRESHOLD = 3  # Threshold for determining missing chapters. Do disable overwriting existing chapters, set to 99999999 ;)
ABS_HOST = ""  # AudiobookShelf Host URL
LIBRARY_ID = ""  # Library UUID
API_KEY = ""  # API Key from user settings
PROVIDER = "audible.de"  # Metadata provider (See available providers in API documentation https://api.audiobookshelf.org/#metadata-providers)
REGION = "DE"  # Region code (e.g., US, DE)
DISABLE_RATE_PROTECTION = False  # Rate protection, disable to speed up but risk timeouts
DISABLE_QUICK_MATCH_NO_ASIN = True  # Disable quick match for books without ASIN to try to find ASIN
USE_TRACKS_AS_CHAPTERS = False  # Use tracks as chapters if no asin available (Fallback)


############################################################################################################

book_info = {}

# Fetch all library items
library_url = f"{ABS_HOST}/api/libraries/{LIBRARY_ID}/items?token={API_KEY}"
print(f"Fetching library items from: {library_url}")
response = requests.get(library_url)

if response.status_code != 200:
    print("Error fetching library items:", response.status_code)
    exit()

items = response.json().get('results', [])
print(f"Found {len(items)} items in the library.")

# Process each item in the library
for item in items:
    book_id = item['id']
    metadata = item['media']['metadata']
    title = metadata.get('title', "Unknown Title")
    subtitle = metadata.get('subtitle', "")
    authors = metadata.get('authorName', "Unknown Author")
    book_info[book_id] = { 'id': book_id, 'title': title, 'status': 'ERROR', 'comment': 'Unknown Error', 'asin': 'N/A' }

    print(f"\n--- Processing Book: {title} ---")

    # Check if book has ASIN (Amazon Standard Identification Number)
    if 'asin' not in metadata or metadata['asin'] is None:
        book_info[book_id]['status'] = 'NO_ASIN'
        book_info[book_id]['asin'] = 'N/A'
        if not DISABLE_QUICK_MATCH_NO_ASIN:
            post_data = {
                "provider": PROVIDER,
                "title": title,
                "author": authors,
            }

            match_url = f"{ABS_HOST}/api/items/{book_id}/match?token={API_KEY}"
            new_item_response = requests.post(match_url, json=post_data)

            if new_item_response.status_code != 200:
                print(f"Error matching book '{title}':", new_item_response.json())
                continue

            new_item = new_item_response.json()
            if 'error' in new_item or 'libraryItem' not in new_item:
                print(f"Error matching book '{title}' (No library item returned).")
                book_info[book_id]['comment'] = 'Asin retrieval failed'
                continue

            if new_item['libraryItem']['media']['metadata'].get('asin') is None:
                print(f"Error matching book '{title}' (No ASIN found).")
                book_info[book_id]['comment'] = 'Asin retrieval failed'
                continue

            book_info[book_id]['comment'] = 'Asin quick matched'

            item = new_item['libraryItem']

    # Check if asin is now available
    if 'asin' not in item['media']['metadata'] or item['media']['metadata']['asin'] is None:
        if USE_TRACKS_AS_CHAPTERS:
            # Get the tracks and use them as chapters
            book_response = requests.get(f"{ABS_HOST}/api/items/{book_id}?expanded=1&token={API_KEY}")
            book_info[book_id]['status'] = 'TRACKS'
            if book_response.status_code != 200:
                print(f"Error fetching book '{title}':", book_response.status_code)
                book_info[book_id]['comment'] = 'Tracks retrieval failed'
                continue
            if len(book_response.json()['media'].get('audioFiles', [])) > 1:
                print(f"Using tracks as chapters for '{title}'.")
                tracks = book_response.json()['media'].get('audioFiles', [])
                new_chapters = []

                current_time = 0
                length_of_book = item['media']['duration']
                for i, track in enumerate(tracks):
                    duration = track['duration']
                    new_chapters.append({
                        "id": i,
                        "start": current_time,
                        "end": current_time + duration,
                        "title": f"Chapter {i+1}",
                        "error": None
                    })
                    current_time += duration - 0.001 # Subtract a tiny amount to avoid rounding issues
                    if length_of_book < current_time:
                        break

                print(f"Chapters updated successfully for '{title}' (Using tracks!).")
                book_info[book_id]['comment'] = 'Tracks used as chapters'
                book_info[book_id]['status'] = 'FINISHED'
                continue
            else:
                print(f"Error using tracks as chapters for '{title}' (No or 1 track found).")
                book_info[book_id]['comment'] = 'Tracks retrieval failed'
        else:
            print(f"Skipping book '{title}' (No ASIN found and Tracks not used as source).")
            book_info[book_id]['comment'] = 'Asin retrieval failed'
            continue
    else:
        book_info[book_id]['asin'] = item['media']['metadata']['asin']

    # Fetch chapters using ASIN
    asin = item['media']['metadata']['asin']
    chapter_url = f"{ABS_HOST}/api/search/chapters?asin={asin}&region={REGION}&token={API_KEY}"
    chapter_response = requests.get(chapter_url)

    if chapter_response.status_code != 200:
        book_info[book_id]['comment'] = 'Chapters retrieval failed'
        print(f"Error fetching chapters for '{title}'. Response Code:", chapter_response.status_code)
        continue

    chapters = chapter_response.json().get('chapters', [])
    print(f"Chapters found for '{title}': {len(chapters)}")

    # Compare current and found chapters
    current_chapters_num = item['media'].get('numChapters', 0)
    found_chapters_num = len(chapters)
    print(f"Current chapter count: {current_chapters_num}")

    if abs(current_chapters_num - found_chapters_num) > CHAPTER_THRESHOLD or current_chapters_num == 0:
        print(f"Chapters are missing or incorrect for '{title}'. Updating...")

        new_chapters = []
        length_of_book = item['media']['duration']  # Book length in seconds

        for i, chapter in enumerate(chapters):
            start = chapter['startOffsetMs'] / 1000
            end = min((chapter['startOffsetMs'] + chapter['lengthMs']) / 1000, length_of_book)

            if start >= length_of_book:
                break

            new_chapters.append({
                "id": i,
                "start": start,
                "end": end,
                "title": chapter['title'],
                "error": None
            })

        # Update chapters for the book
        update_url = f"{ABS_HOST}/api/items/{book_id}/chapters?token={API_KEY}"
        update_data = {"chapters": new_chapters}
        update_response = requests.post(update_url, json=update_data)

        if update_response.status_code == 200:
            print(f"Chapters updated successfully for '{title}'.")
            book_info[book_id]['status'] = 'FINISHED'
            book_info[book_id]['comment'] = 'Chapters updated'
        else:
            print(f"Error updating chapters for '{title}'. Response Code:", update_response.status_code)
            book_info[book_id]['comment'] = 'Chapters update failed'
    else:
        book_info[book_id]['status'] = 'FINISHED'
        book_info[book_id]['comment'] = 'No chapters to update'
        print(f"Chapters are fine for '{title}'.")

    # Respect rate-limiting if enabled
    if not DISABLE_RATE_PROTECTION:
        time.sleep(2)

print("\n--- Summary ---")
for book_id, info in book_info.items():
    print(f"{info['title']} ({info['status']}): {info['comment']}\nLink: {ABS_HOST}/item/{book_id}\n")
    print("-" * 50)

print("\n--- Failed Books ---")
for book_id, info in book_info.items():
    if info['status'] == 'ERROR':
        print(f"{info['title']} ({info['status']}): {info['comment']}\nLink: {ABS_HOST}/item/{book_id}\n")
        print("-" * 50)

File: web/src/features/tools/quickMatchChapters.ts
import { useApi } from "@/shared/composables/useApi";
import type { ToolResult } from "./types";
import {
  AbsBookLibraryItemExpanded,
  AbsBookLibraryItemMinified,
  AbsLibrary,
  AbsSearchChaptersData,
} from "@vito0912/abs-ts-sdk";
import { AbsLibraryItemChaptersUpdateRequest } from "@vito0912/abs-ts-sdk/types";

const { absClient, addLog } = useApi();

function delay(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function updateBookChapters(
  bookId: string,
  chapters: AbsLibraryItemChaptersUpdateRequest,
) {
  try {
    await absClient.libraryItems.updateChapters(bookId, chapters);
    return true;
  } catch (error) {
    console.error(`Error updating chapters for book ${bookId}:`, error);
    return false;
  }
}

function createChaptersFromTracks(tracks: any[], bookDuration: number) {
  const chapters = [];
  let currentTime = 0;

  for (let i = 0; i < tracks.length; i++) {
    const track = tracks[i];
    const duration = track.duration;
    const trackTitle = track.metadata.filename.split(".")[0];

    chapters.push({
      id: i,
      start: currentTime,
      end: currentTime + duration,
      title: trackTitle,
      error: null,
    });

    currentTime += duration - 0.001; // Subtract tiny amount to avoid rounding issues
    if (bookDuration < currentTime) {
      break;
    }
  }

  return chapters;
}

function createChaptersFromAsin(chapters: any[], bookDuration: number) {
  const newChapters = [];

  for (let i = 0; i < chapters.length; i++) {
    const chapter = chapters[i];
    const start = chapter.startOffsetMs / 1000;
    const end = Math.min(
      (chapter.startOffsetMs + chapter.lengthMs) / 1000,
      bookDuration,
    );

    if (start >= bookDuration) {
      break;
    }

    newChapters.push({
      id: i,
      start: start,
      end: end,
      title: chapter.title,
      error: null,
    });
  }

  return newChapters;
}

export async function executeMatchAudiobookChapters(
  formData: Record<string, any>,
): Promise<ToolResult> {
  try {
    const {
      chapterThreshold = 3,
      libraryId,
      provider = "audible.com",
      region = "US",
      disableRateProtection = false,
      searchForAsin = true,
      useTracksAsChapters = false,
    } = formData;

    const bookInfo: Record<string, any> = {};

    addLog(`Fetching library items from library: ${libraryId}`);

    const library = await absClient.libraries.getById(libraryId);
    if (!library || (library as AbsLibrary).mediaType !== "book") {
      throw new Error(`Library with ID ${libraryId} not a valid book library.`);
    }

    const items = ((await absClient.libraries.listItems(libraryId)).results ||
      []) as AbsBookLibraryItemMinified[];
    addLog(`Found ${items.length} items in the library.`);

    // Process each item in the library
    for (const item of items) {
      const bookId = item.id;
      const metadata = item.media.metadata;
      const title = metadata.title || "Unknown Title";
      const authors = metadata.authorName || "Unknown Author";

      bookInfo[bookId] = {
        id: bookId,
        title: title,
        status: "ERROR",
        comment: "Unknown Error",
        asin: "N/A",
      };

      addLog(`\n--- Processing Book: ${title} ---`);

      // Check if book has ASIN
      if (!metadata.asin) {
        bookInfo[bookId].status = "NO_ASIN";
        bookInfo[bookId].asin = "N/A";

        if (searchForAsin) {
          addLog(`Searching for ASIN for "${title}"...`);
          const searchResults = await absClient.search.getBooks({
            title,
            author: authors,
            provider,
          });

          if (searchResults.length === 0) {
            addLog(`Error matching book "${title}" (No results found).`);
            bookInfo[bookId].comment = "ASIN retrieval failed";
            continue;
          }

          const bestMatch = searchResults[0];
          const asin = bestMatch.asin;

          if (!asin) {
            addLog(`Error matching book "${title}" (No ASIN found).`);
            bookInfo[bookId].comment = "ASIN retrieval failed - No ASIN found";
            continue;
          }

          item.media.metadata.asin = asin;
          metadata.asin = asin;
          addLog(`ASIN found: ${asin}`);
        }
      }

      // Check if ASIN is now available
      if (!item.media.metadata.asin) {
        if (useTracksAsChapters) {
          addLog(`Using tracks as chapters for "${title}".`);
          bookInfo[bookId].status = "TRACKS";

          const bookDetails: AbsBookLibraryItemExpanded =
            (await absClient.libraryItems.getById(bookId, {
              expanded: true,
            })) as AbsBookLibraryItemExpanded;
          if (!bookDetails) {
            addLog(
              `Error fetching book "${title}": Failed to get book details`,
            );
            bookInfo[bookId].comment = "Tracks retrieval failed";
            continue;
          }

          const audioFiles = bookDetails.media.audioFiles || [];
          if (audioFiles.length <= 1) {
            addLog(
              `Error using tracks as chapters for "${title}" (No or 1 track found).`,
            );
            bookInfo[bookId].comment = "Tracks retrieval failed";
            continue;
          }

          const currentChaptersNum = item.media.numChapters || 0;
          const tracksNum = audioFiles.length;

          if (
            currentChaptersNum === 0 ||
            Math.abs(currentChaptersNum - tracksNum) > chapterThreshold
          ) {
            addLog(
              `Chapters are missing or incorrect for "${title}" (Current num: ${currentChaptersNum}, Tracks num: ${tracksNum}). Updating...`,
            );

            const newChapters = createChaptersFromTracks(
              audioFiles,
              item.media.duration,
            );
            const success = await updateBookChapters(bookId, {
              chapters: newChapters,
            });

            if (success) {
              addLog(
                `Chapters updated successfully for "${title}" (Using tracks!).`,
              );
              bookInfo[bookId].comment = "Tracks used as chapters";
              bookInfo[bookId].status = "FINISHED";
            } else {
              addLog(`Error updating chapters for "${title}".`);
              bookInfo[bookId].comment = "Chapters update failed";
            }
          } else {
            addLog(`Chapters are fine for "${title}".`);
            bookInfo[bookId].status = "FINISHED";
            bookInfo[bookId].comment = "No chapters to update";
          }
          continue;
        } else {
          addLog(
            `Skipping book "${title}" (No ASIN found and Tracks not used as source).`,
          );
          bookInfo[bookId].comment = "ASIN retrieval failed";
          continue;
        }
      } else {
        bookInfo[bookId].asin = item.media.metadata.asin;
      }

      // Fetch chapters using ASIN
      const asin = item.media.metadata.asin;
      addLog(`Fetching chapters for ASIN: ${asin}`);

      const chapters = (
        (await absClient.search.getChapters({
          asin: asin,
          region: region,
        })) as AbsSearchChaptersData
      ).chapters;

      if (chapters.length === 0) {
        bookInfo[bookId].comment = "No chapters found";
        addLog(`No chapters found for "${title}".`);
        continue;
      }

      addLog(`Chapters found for "${title}": ${chapters.length}`);

      // Compare current and found chapters
      const currentChaptersNum = item.media.numChapters || 0;
      const foundChaptersNum = chapters.length;
      addLog(`Current chapter count: ${currentChaptersNum}`);

      if (
        Math.abs(currentChaptersNum - foundChaptersNum) > chapterThreshold ||
        currentChaptersNum === 0
      ) {
        addLog(`Chapters are missing or incorrect for "${title}". Updating...`);

        const newChapters = createChaptersFromAsin(
          chapters,
          item.media.duration,
        );
        const success = await updateBookChapters(bookId, {
          chapters: newChapters,
        });

        if (success) {
          addLog(`Chapters updated successfully for "${title}".`);
          bookInfo[bookId].status = "FINISHED";
          bookInfo[bookId].comment = "Chapters updated";
        } else {
          addLog(`Error updating chapters for "${title}".`);
          bookInfo[bookId].comment = "Chapters update failed";
        }
      } else {
        bookInfo[bookId].status = "FINISHED";
        bookInfo[bookId].comment = "No chapters to update";
        addLog(`Chapters are fine for "${title}".`);
      }

      if (!disableRateProtection) {
        await delay(2000);
      }
    }

    addLog("\n--- Summary ---");
    for (const [_bookId, info] of Object.entries(bookInfo)) {
      addLog(`${info.title} (${info.status}): ${info.comment}`);
      addLog("-".repeat(50));
    }

    addLog("\n--- Failed Books ---");
    for (const [_bookId, info] of Object.entries(bookInfo)) {
      if (info.status !== "FINISHED") {
        addLog(`${info.title} (${info.status}): ${info.comment}`);
        addLog("-".repeat(50));
      }
    }

    return {
      success: true,
      message: "Audiobook chapters matching completed",
      timestamp: new Date().toISOString(),
    };
  } catch (error: any) {
    return {
      success: false,
      message: "Failed to match audiobook chapters",
      error: error.message || "Unknown error",
      timestamp: new Date().toISOString(),
    };
  }
}
